lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -665,14 +688,159 @@ def write(chunk)
+    # cause of the first failed log entry. For example, if we switched the order
+    # of the log entries, then we would get:
+    # {
+    #    "error" : {
+    #       "code" : 400,
+    #       "message" : "Log name contains illegal character :",
+    #       "status" : "INVALID_ARGUMENT",
+    #       "details": ...
+    #    }
+    # }
+    # We will ignore it anyway and look at the details instead which includes
+    # info for all failed log entries.
+    #
+    # In this example, the logEntryErrors we are extracting should be:
+    # {
+    #   "0": {
+    #     "code": 7,
+    #     "message": "User not authorized."
+    #   },
+    #   "1": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   },
+    #   "3": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   }
+    # }
+    def extract_log_entry_errors(error)
+      error_body = ensure_hash(JSON.parse(error.body))
+      error = ensure_hash(error_body['error'])
+      error_details = ensure_array(error['details'])
+      partial_errors = error_details.detect do |error_detail|
+        ensure_hash(error_detail)['@type'] == PARTIAL_ERROR_FIELD
+      end
+      unless partial_errors
+        @log.warn 'Failed to detect an error detail with type' \
+          " #{PARTIAL_ERROR_FIELD}."
+      end
+      ensure_hash(ensure_hash(partial_errors)['logEntryErrors'])
+    rescue JSON::ParserError => e
+      @log.warn 'Failed to extract log entry errors from the error details' \
+                " because of JSON parsing error: #{error.body}.", error: e
+      {}
+    end
+
+    # Given the logEntryErrors, construct a map from errors to a list of indexes
+    # of log entries that failed for this specific error.
+    #
+    # The keys in this map are [error_code, error_message] pairs, and the values
+    # are a list of stringified indexes of log entries that failed due to this
+    # error.
+    #
+    # For logEntryErrors as in the example above (in the header comment for
+    # extract_log_entry_errors), the constructed map would be:
+    # {
+    #   [7, 'User not authorized.']: ['0'],
+    #   [3, 'Log name contains illegal character :']: ['1', '3']
+    # }
+    def construct_error_details_map(log_entry_errors)
+      error_details_map = Hash.new { |h, k| h[k] = [] }
+      log_entry_errors.each do |index, error|
+        error_hash = ensure_hash(error)
+        unless error_hash['code'] && error_hash['message']
+          @log.warn "Entry with index #{index} is missing 'code' or 'message'" \
+                    ' field.'
+          return {}
+        end
+        error_key = [error_hash['code'], error_hash['message']].freeze
+        # TODO(qingling128): Convert indexes to integers.
+        error_details_map[error_key] << index
+      end
+      error_details_map
+    end
+
+    def ensure_array(value)
+      ensure_non_empty(Array.try_convert(value))
```

Comment:
```ruby
Array.try_convert(value) or fail JSON::ParserError
```
Then you don't need `ensure_non_empty`. Same for `ensure_hash`:
```ruby
Hash.try_convert(value) or fail JSON::ParserError
```

Reply:
Ah. I tried this already. Rubocop does not like this type of syntax.

Reply:
This seems to fix it:
```diff
--- a/lib/fluent/plugin/out_google_cloud.rb
+++ b/lib/fluent/plugin/out_google_cloud.rb
@@ -825,16 +825,11 @@ module Fluent
     end
 
     def ensure_array(value)
-      ensure_non_empty(Array.try_convert(value))
+      Array.try_convert(value) || (fail JSON::ParserError, "#{value.class}")
     end
 
     def ensure_hash(value)
-      ensure_non_empty(Hash.try_convert(value))
-    end
-
-    def ensure_non_empty(converted_value)
-      fail JSON::ParserError unless converted_value
-      converted_value
+      Hash.try_convert(value) || (fail JSON::ParserError, "#{value.class}")
     end
 
     private
```